### ISSUE TITLE
Implement GraphlQl Error Handling: Multiple Resolvers

### ISSUE THREAD

**Issue by kb-0311** [PalisadoesFoundation/talawa-api#1486]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is.
As mentioned, I must work on implementing schema changes in the API.
the resolvers I will make changes to 5 resolvers

I will need a few volunteers to make appropriate changes to the admin and mobile apps.
```
createMember(input: UserAndOrganizationInput!): Organization! @auth

    createAdmin(data: UserAndOrganizationInput!): User!
      @auth
      @role(requires: SUPERADMIN)

    createComment(postId: ID!, data: CommentInput!): Comment @auth

    createDirectChat(data: createChatInput!): DirectChat! @auth

    createDonation(
      userId: ID!
      orgId: ID!
      payPalId: ID!
      nameOfUser: String!
      amount: Float!
      nameOfOrg: String!
    ): Donation!
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Approach to be followed (optional)**
A clear and concise description of approach to be followed.

**Additional context**

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by kb-0311** [comment]
assign this to me @EshaanAgg

**Comment by Min2who** [comment]
Can i Work with you @kb-0311

**Comment by kb-0311** [comment]
@Min2who Would you like to make a sister PR to the talawa-admin or mobile app for the changes I am about to make sure?

**Comment by Min2who** [comment]
> @Min2who Would you like to make a sister PR to the talawa-admin or mobile app for the changes I am about to make sure?

Certainly ! Go ahead with your PR

**Comment by Cioppolo14** [comment]
@Min2who  Please create a seperate issue for your contribution.

**Comment by Min2who** [comment]
> @Min2who Please create a seperate issue for your contribution.

ok making it in the talawa-admin

**Comment by NamitBhutani** [comment]
If no one is working on this issue, I'd like to work on it! @EshaanAgg

**Comment by Cioppolo14** [comment]
@kb-0311 Are you still working on this?

**Comment by kb-0311** [comment]
@Cioppolo14 Nope, I will unassign myself and then someone else can collaborate with @Min2who on this issue.

**Comment by Cioppolo14** [comment]
@NamitBhutani I have assigned you, you now have two issues.

**Comment by NamitBhutani** [comment]
@Min2who can you please update me as to what is the status of your PR in talawa-admin? And please link the PR too!
Thanks :D

**Comment by Min2who** [comment]
> @Min2who can you please update me as to what is the status of your PR in talawa-admin? And please link the PR too! Thanks :D

@NamitBhutani it is up for you to work on it first

**Comment by NamitBhutani** [comment]
> > @Min2who can you please update me as to what is the status of your PR in talawa-admin? And please link the PR too! Thanks :D
> 
> @NamitBhutani it is up for you to work on it first

Ok! I am working on it :D

**Comment by NamitBhutani** [comment]
@palisadoes I am working on the error handling for the resolvers, but I am curious as to is there any standard method that is followed. My current approach is to use try catch blocks across all the await functions to catch any errors that occur during the execution. Is there a better method for this too?

**Comment by palisadoes** [comment]
@kb-0311 @xoldyckk I know you had many discussions about this, but it was never documented beyond this discussion
- https://github.com/PalisadoesFoundation/talawa-api/discussions/1214

@NamitBhutani take a look at this discussion thread. We'll need a markdown file from you on the final outcome so that we can document it once and for all. Please take input from the others, but don't let that prevent you from starting.

**Comment by kb-0311** [comment]
@NamitBhutani I had written a guide regarding the "errors-in-schema" approach in the API. You can check out this file here all of your questions are answered there :

https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/docs/developers/talawa-api/error-handling.md

**Comment by NamitBhutani** [comment]
> @NamitBhutani I had written a guide regarding the "errors-in-schema" approach in the API. You can check out this file here all of your questions are answered there including the structure and examples:
> 
> https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/docs/developers/talawa-api/error-handling.md

I have read through it, but I had an issue, GraphQL doesn't allow unions to include interfaces as "technically" they are not objects. It's not an issue as we can always make a default type that implements the interface, but just wanted to point it out as it is mentioned in doc.

**Comment by NamitBhutani** [comment]
Hi! @palisadoes I have made changes to the way error handling works for one of the resolver. The commit does not completely fix the issue ( as only 1 resolver is addressed). I wanted to raise a PR so that we can discuss on the approach and then I can proceed accordingly for other resolvers. Am I allowed to open a PR in such a case? Or is there any better way you can suggest?

**Comment by palisadoes** [comment]
Ask your questions here. The persons who defined the policy are on this thread.

**Comment by NamitBhutani** [comment]
For the createMember mutation resolver, I have modelled the errors as follows `type OrganizationNotFoundError implements UserErrorInterface {
    message: String!
    code: String!
    param: String!
  }

  type MemberAlreadyinOrganizationError implements UserErrorInterface {
    message: String!
    code: String!
    param: String!
  }

  type UserNotFoundError implements UserErrorInterface {
    message: String!
    code: String!
    param: String!
  }` ,  the interface being ` interface UserErrorInterface {
    message: String!
    code: String!
    param: String!
  }`. Instead of the approach mentioned in https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/docs/developers/talawa-api/error-handling.md, I have gone through with a slightly different route, i.e. having `union CreateMemberResult =
      Organization
    | OrganizationNotFoundError
    | MemberAlreadyinOrganizationError
    | UserNotFoundError` as the return type of the mutation `createMember(input: UserAndOrganizationInput!): CreateMemberResult! @auth`, where `Organization` is the data it must return if there is no error.
One example of it being used in the resolver function code is `if (!organization) {return 
{
      __typename: "OrganizationNotFoundError",
      message: requestContext.translate(
        ORGANIZATION_NOT_FOUND_ERROR.MESSAGE
      ) as string,
      code: ORGANIZATION_NOT_FOUND_ERROR.CODE,
      param: ORGANIZATION_NOT_FOUND_ERROR.PARAM,
    };
  }`. I wanted @palisadoes @kb-0311 @xoldyckk to have a look at this approach and if there are problems we can discuss them , else I'll move forward with implementing it to other resolvers too.

**Comment by NamitBhutani** [comment]
> For the `createMember` mutation resolver, I have modelled the errors as follows
> 
> ```
> type OrganizationNotFoundError implements UserErrorInterface {
>     message: String!
>     code: String!
>     param: String!
>   }
> ```
> 
> ```
> type MemberAlreadyinOrganizationError implements UserErrorInterface {
>    message: String!
>    code: String!
>    param: String!
>  }
> ```
> 
> ```
> type UserNotFoundError implements UserErrorInterface {
>   message: String!
>   code: String!
>   param: String!
> }  
> ```
> 
> The interface being
> 
> ```
> interface UserErrorInterface {
>     message: String!
>     code: String!
>     param: String!
>   }
> ```
> 
> Instead of the approach mentioned in https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/docs/developers/talawa-api/error-handling.md, I have gone through with a slightly different route, i.e. having
> 
> ```
> union CreateMemberResult =
>       Organization
>     | OrganizationNotFoundError
>     | MemberAlreadyinOrganizationError
>     | UserNotFoundError 
> ```
> 
> as the return type of the mutation `createMember(input: UserAndOrganizationInput!): CreateMemberResult! @auth ` where `Organization` is the data it must return if there is no error. One example of it being used in the resolver function code is
> 
> ```
> if (!organization) {return 
> {
>       __typename: "OrganizationNotFoundError",
>       message: requestContext.translate(
>         ORGANIZATION_NOT_FOUND_ERROR.MESSAGE
>       ) as string,
>       code: ORGANIZATION_NOT_FOUND_ERROR.CODE,
>       param: ORGANIZATION_NOT_FOUND_ERROR.PARAM,
>     };
>   }
> ```
> 
> Also attaching the [resolver file](https://pastebin.com/VyHq0jxm) I wanted @palisadoes @kb-0311 @xoldyckk to have a look at this approach and if there are problems we can discuss them , else I'll move forward with implementing it to other resolvers too.

Is there any update on it? So that I can continue working on it!

**Comment by xoldyckk** [comment]
@NamitBhutani the implementation must be stage 6a as listed in this article:- https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors

nodes and edges in the schema are usually not designed with errors in schema approach because they rarely make use of manual user input, but it is still doable, take a look at this example:- https://oliverjumpertz.com/blog/the-ultimate-guide-to-graphql-error-handling/

https://github.com/PalisadoesFoundation/talawa-api/pull/1585#discussion_r1441534417
👆 here i've explained how the mutations should be designed, for errors you have to add an extra field named `clientErrors` or `userErrors` or `errors`, rest is mentioned in the article I linked at the very start

a special case is resolver argument validation, it should have a predictable and stable structure so client developers can easily make use of the errors to provide feedback to the end user, i recommend a flatmap:-
```graphql
interface Error {
  message: String!
}

# reusable across many mutationresolvers
type InvalidArgumentError implements Error {
  message: String!
  path: [String!]!
}

# notice the absense of path field in this error type, when using errors in schema approach, argument errors are the # # only places where path field makes sense 
type SomeOtherError implements Error {
  message: String!
  someOtherField: SomeOtherType!
}

union CreateUserError = InvalidArgumentError | SomeOtherError

type CreateUserPayload {
  clientErrors: [CreateUserError!]!
  user: User
}

type Mutation {
  createUser(input: CreateUserInput!): CreateUserPayload
}
```

for an input object like this:-
```typescript
type Input = {
  email: string;
  location: {
    latitude: string
  };
  urls: string[]
}
```
here's the example of flatmap of argument errors(for example here all arguments are invalid):-
```typescript
const clientErrors = [
  {
    __typename: "InvalidArgumentError",
    message: "This email is already registered with our application.",
    path: ["input", "email"],
  },
  {
    __typename: "InvalidArgumentError",
    message: "This is not a valid latitude.",
    path: ["input", "location", "latitude"],
  },
  {
    __typename: "InvalidArgumentError",
    message: "This is not a valid url.",
    path: ["input", "urls", "0"],
  },
  {
    __typename: "InvalidArgumentError",
    message: "This is not a valid url.",
    path: ["input", "urls", "10"],
  }
]
```
```typescript
const clientErrors = [
  {
    __typename: "SomeOtherError",
    message: "Some error message.",
    someOtherField: ...
  },
]
```

**Comment by xoldyckk** [comment]
One more thing, current talawa-api's
```typescript
{
   message: String!
   code: String!
   param: String!
}
```
error handling practice is outdated, code and param are useless when using errors in schema approach, because that information is contained within the `__typename` field of schema and also because the errors are localized to places where they occur.

the only usable part is `message` which can be used to provide localized error messages(l18n, i18n), I believe some modifications will have to be done to way those error message are returned, probably would require creation of new generic classes that can be used to construct error response types

**Comment by NamitBhutani** [comment]
I have made a PR @xoldyckk with changes to the createMember resolver. Please review once and let me know if there any changes, so that I follow the same approach then after the final changes you recommend for other resolvers too!

**Comment by NamitBhutani** [comment]
I have found a bug in createMember resolver, that is adding a member to the organization even after the member is already in the organization, this bug somehow made past the tests and is present in the main production branch.
The issue is I have tried to fix it but to no use , I would like to explain it to senior maintainers so that we can fix it. This bug is also blocking the PR for this issue now. @palisadoes
